### ISSUE TITLE
"Length cannot be less than zero" at Shouldly.ShouldlyMessage.ToString()

### ISSUE THREAD

**Issue by vorou** [shouldly/shouldly#105]
This assertion:

```
actual.ShouldBeEmpty();
```

results into this output:

```
Test 'Overseer.Tests.TenderRepositoryTests.GetMostExpensive_OneTenderWasPublishedYesterday_SkipsIt' failed: System.ArgumentOutOfRangeException : Length cannot be less than zero.
Parameter name: length
	at System.String.Substring(Int32 startIndex, Int32 length)
	at Shouldly.ShouldlyMessageGenerator.GetCodePart(TestEnvironment environment)
	at Shouldly.ShouldBeEmptyMessageGenerator.GenerateErrorMessage(TestEnvironment environment, Object actual)
	at Shouldly.ShouldlyMessage.GenerateShouldMessage(Object expected)
	at Shouldly.ShouldlyMessage.ToString()
	at Shouldly.ShouldBeEnumerableTestExtensions.ShouldBeEmpty[T](IEnumerable`1 actual)
	TenderRepositoryTests.cs(105,0): at Overseer.Tests.TenderRepositoryTests.GetMostExpensive_OneTenderWasPublishedYesterday_SkipsIt()
```

Here's how I get the `actual` value (`elastic` is an instance of [NEST](https://github.com/Mpdreamz/NEST) client):

```
return elastic.Search<Tender>(q => q.MatchAll().SortDescending(t => t.TotalPrice)).Documents.Take(limit);
```

and it's not empty, so I actually expect the assertion to fail, but with more elaborate message.

**Comment by asgerhallas** [comment]
I have this too, but only on my build server:

System.ArgumentOutOfRangeException : Length cannot be less than zero.
    Parameter name: length
       at System.String.Substring(Int32 startIndex, Int32 length)
       at Shouldly.ShouldlyMessageGenerator.GetCodePart(TestEnvironment environment) in c:\TeamCity\buildAgent\work\10efaabfa8adbd4e\src\Shouldly\ShouldlyMessageGenerators.cs:line 100
       at Shouldly.DefaultMessageGenerator.GenerateErrorMessage(TestEnvironment environment, Object expected) in c:\TeamCity\buildAgent\work\10efaabfa8adbd4e\src\Shouldly\ShouldlyMessageGenerators.cs:line 76
       at Shouldly.ShouldlyMessage.GenerateShouldMessage(Object expected) in c:\TeamCity\buildAgent\work\10efaabfa8adbd4e\src\Shouldly\ShouldlyMessage.cs:line 60
       at Shouldly.ShouldlyMessage.ToString() in c:\TeamCity\buildAgent\work\10efaabfa8adbd4e\src\Shouldly\ShouldlyMessage.cs:line 32
       at Shouldly.Should.Throw[TException](Action actual) in c:\TeamCity\buildAgent\work\10efaabfa8adbd4e\src\Shouldly\ShouldThrowExtensions.cs:line 25
       at Energy10.Tests.Infrastructure.Persistence.ResolveReferencesTests.FailsWhenReferencingSameInstanceTwiceAndItIsNotParentOrRoot() in c:\BuildAgent\work\881cef7e85e37837\Src\Energy10.Tests\Infrastructure\Persistence\ResolveReferencesTests.cs:line 94

And only this single test, and it is asserting with `Should.Throw` and not `ShouldBeEmpty`.

**Comment by JakeGinnivan** [comment]
Is the source code available or are you building without PDB's?

**Comment by asgerhallas** [comment]
Ah, yes I see, it's only on release builds with no pdb's or source.

**Comment by asgerhallas** [comment]
I'll try to deploy the pdb's too and give you feedback in a second :)

**Comment by JakeGinnivan** [comment]
We need to gracefully fall back. I thought that I had done that. 

Opened #170 to track the issue

**Comment by asgerhallas** [comment]
Hmm... actually the PDB's a being built and the source code is available too.

**Comment by JakeGinnivan** [comment]
Damn.. Not sure what the issue is then. Should be able to craft a fix anyways so we do not crash.

**Comment by asgerhallas** [comment]
Not sure either, but it consistently fails on Release builds now. Also locally, in Resharper runner and for any type of assertion it seems. 

But great if it could suppress the error somehow... :)

**Comment by chaitanyagurrapu** [comment]
Would you be able to update your version of Shouldly and then post the updated stack trace please?

I am having trouble reproducing this with the current version. I compiled a sandbox nunit test project in release mode. In the project's advanced build settings, changed the "Debug Info" option to "None". Copied the dll (minus the pdb) to a completely different machine - a windows 2008 server VM. And used the nunit gui runner to run it there. You can see below what I see. 

![shouldlynodebug](https://cloud.githubusercontent.com/assets/853059/5815547/4b33091e-a0e2-11e4-8ece-f4904f7431d3.png)


![shouldlynofail](https://cloud.githubusercontent.com/assets/853059/5815469/903206ce-a0e1-11e4-9700-eec80d942da2.png)

**Comment by asgerhallas** [comment]
Sorry for the delayed answer. I've just fetched the latest Shouldly source from github and implemented this simple test:

    [TestFixture]
    public class ShouldFail
    {
        [Test]
        public void Hello()
        {
            "asger".ShouldBe("horse");
        }
    }

Built with Release config and it fails with: 

    System.ArgumentOutOfRangeException : Length cannot be less than zero.
Parameter name: length
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Shouldly.ShouldlyAssertionContext.GetCodePart() in ShouldlyAssertionContext.cs: line 90
   at Shouldly.ShouldlyAssertionContext..ctor(Object expected, Object actual) in ShouldlyAssertionContext.cs: line 76
   at Shouldly.ExpectedActualShouldlyMessage..ctor(Object expected, Object actual) in ShouldlyMessage.cs: line 25
   at Shouldly.ShouldlyCoreExtensions.AssertAwesomely(T actual, Func`2 specifiedConstraint, Object originalActual, Object originalExpected) in ShouldlyCoreExtensions.cs: line 18
   at Shouldly.ShouldBeTestExtensions.ShouldBe(T actual, T expected) in ShouldBeTestExtensions.cs: line 16
   at Shouldly.Tests.ShouldFail.Hello() in ShouldCompleteInTests.cs: line 14

**Comment by JakeGinnivan** [comment]
What version of visual studio/.net

**Comment by asgerhallas** [comment]
VS 2013 Premium update 3

**Comment by asgerhallas** [comment]
It skips too many line numbers when it reads the code lines from the file. It finds the file alright though, but this is the content of codeLines for the above test in a single file:

            }
        }
    }

**Comment by asgerhallas** [comment]
LineNumber for Release config is 11 and for Debug config it is 10.

**Comment by asgerhallas** [comment]
I don't know if it's always the case, but the -1 in this line

           LineNumber = originatingFrame.GetFileLineNumber() - 1;

seems to be dependent on the build config.

**Comment by asgerhallas** [comment]
I thought I'd make a pull request for you, but the offset of line number is really not consistent. My simple test was two lines offset, but quite a few other tests in Shouldly.Tests failed for me when run in Release. I'm sorry, I don't see a good solution right now.

**Comment by JakeGinnivan** [comment]
I've put up a pull request. The test still fail in release mode, but because the error messages don't match. Not the crash

**Comment by JakeGinnivan** [comment]
This has been fixed in 2.4.0

**Comment by asgerhallas** [comment]
Sorry for the late reply. Thanks for the fix, I wish I had the time to help out more!
